tensorcircuit/mpscircuit.py

```diff
@@ -225,35 +298,224 @@ def apply_double_gate(
+        for i in range(length - 1):
+            nodes[i].get_edge(3) ^ nodes[i + 1].get_edge(0)
+        for i in range(length):
+            output_edges.append(nodes[i].get_edge(1))
+        for i in range(length):
+            output_edges.append(nodes[i].get_edge(2))
+        output_edges.append(nodes[length - 1].get_edge(3))
+        gate = contractor(nodes, output_edge_order=output_edges)
+        return Gate(gate.tensor[0, ..., 0])
+
+    @classmethod
+    def reduce_tensor_dimension(
+        cls,
+        tensor_left: Tensor,
+        tensor_right: Tensor,
+        center_left: bool = True,
+        **rules: Any,
+    ) -> Tuple[Tensor, Tensor]:
+        """
+        Reduce the bond dimension between two general tensors by SVD
+        """
+        ni = tensor_left.shape[0]
+        nk = tensor_right.shape[-1]
+        T = backend.einsum("iaj,jbk->iabk", tensor_left, tensor_right)
+        T = backend.reshape(T, (ni * 2, nk * 2))
+        new_tensor_left, new_tensor_right = split_tensor(
+            T, center_left=center_left, **rules
+        )
+        new_tensor_left = backend.reshape(new_tensor_left, (ni, 2, -1))
+        new_tensor_right = backend.reshape(new_tensor_right, (-1, 2, nk))
+        return new_tensor_left, new_tensor_right
+
+    def reduce_dimension(
+        self,
+        index_left: int,
+        index_right: int,
+        center_left: bool = True,
+    ) -> None:
+        """
+        Reduce the bond dimension between two adjacent sites by SVD
+        """
+        if index_left > index_right:
+            self.reduce_dimension(index_right, index_left, center_left=center_left)
+            return
+        assert index_left + 1 == index_right
+        assert self._mps.center_position in [index_left, index_right]
+        tensor_left = self._mps.tensors[index_left]
+        tensor_right = self._mps.tensors[index_right]
+        new_tensor_left, new_tensor_right = self.reduce_tensor_dimension(
+            tensor_left, tensor_right, center_left=center_left, **self.rules
+        )
+        self._mps.tensors[index_left] = new_tensor_left
+        self._mps.tensors[index_right] = new_tensor_right
+        if center_left:
+            self._mps.center_position = index_left
+        else:
+            self._mps.center_position = index_right
+
+    def apply_MPO(
+        self,
+        tensors: Sequence[Tensor],
+        index_left: int,
+        center_left: bool = True,
+    ) -> None:
+        """
+        Apply a MPO to the MPS
+        """
+        # step 1:
+        #     contract tensor
+        #           a
+        #           |
+        #     i-----O-----j            a
+        #           |        ->        |
+        #           b             ik---X---jl
+        #           |
+        #     k-----T-----l
+        # step 2:
+        #     canonicalize the tensors one by one from end1 to end2
+        # setp 3:
+        #     reduce the bond dimension one by one from end2 to end1
```

Comment:
why we take a round sweep to separate step2 and step3?

Reply:
In MPS, if you reduce the dimension of the bond between site A and site B using SVD, the reduction is optimal (in the sense of fidelity) only when the MPS is canonicalized and the center is at either A or B. The reason is that only in such case the environment density outsite A and B is identity, so the optimal truncation between A and B is also the optical truncation of the whole MPS. In this specific case, the tensors between end1 and end2 are not canonicalized after step 1, so the step 2 is needed to first canonicalize them.

Reply:
thanks for the nice explanation, I recall on these aspects, the optimal compression thing!